### PR TITLE
[Feature][Connector-V2][SelectDB Cloud] Support SelectDB Cloud Sink Connector

### PR DESCRIPTION
**Pull request by zy-kkk**
## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
        SelectDB Cloud is a cloud native database, and docker cannot be used to start test cases.
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

## Test procedure
SelectDB Table Schema
```
CREATE  TABLE `TypeTest` (
    c_map varchar(1000),
    c_array array<int>,
    c_string String,
    c_boolean boolean,
    c_tinyint tinyint,
    c_smallint smallint,
    c_int int,
    c_bigint bigint,
    c_float float,
    c_double double,
    c_decimal  decimal(27, 9),
    c_null String,
    c_bytes String,
    c_date date,
    c_timestamp datetime
) ENGINE=OLAP
DUPLICATE KEY(`c_map`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`c_map`) BUCKETS 4
```
1. ST Engine
   * config
```
   env {
      execution.parallelism = 1
      job.mode = "BATCH"
      checkpoint.interval = 10
    }

      source {
        FakeSource {
          row.num = 10
          parallelism = 1
          schema = {
            fields {
              c_map = "map<string, string>"
              c_array = "array<int>"
              c_string = string
              c_boolean = boolean
              c_tinyint = tinyint
              c_smallint = smallint
              c_int = int
              c_bigint = bigint
              c_float = float
              c_double = double
              c_decimal = "decimal(27, 9)"
              c_null = "null"
              c_bytes = bytes
              c_date = date
              c_timestamp = timestamp
            }
          }
          result_table_name = "fake"
        }
      }
      
      transform {
      }
      
      sink {
        SelectDBCloud {
          load-url="ip:port"
          jdbc-url="ip:port"
          cluster-name="SeaTunnel"
          table.identifier="test.TypeTest"
          username="admin"
          password="******"
          selectdb.config {
              file.type="json"
              file.strip_outer_array="false"
          }
        }
      }
```
  * result
image
image

2. Flink
   * config
```
   env {
        execution.parallelism = 1
        job.mode = "BATCH"
    }

      source {
        FakeSource {
          row.num = 10
          parallelism = 1
          schema = {
            fields {
              c_map = "map<string, string>"
              c_array = "array<int>"
              c_string = string
              c_boolean = boolean
              c_tinyint = tinyint
              c_smallint = smallint
              c_int = int
              c_bigint = bigint
              c_float = float
              c_double = double
              c_decimal = "decimal(27, 9)"
              c_null = "null"
              c_bytes = bytes
              c_date = date
              c_timestamp = timestamp
            }
          }
          result_table_name = "fake"
        }
      }
      
      transform {
      }
      
      sink {
        SelectDBCloud {
          load-url="ip:port"
          jdbc-url="ip:port"
          cluster-name="SeaTunnel"
          table.identifier="test.TypeTest"
          username="admin"
          password="******"
          selectdb.config {
              file.type="json"
              file.strip_outer_array="false"
          }
        }
      }
```
  * result
image
image

3. Spark
    * Config
```
  env {
        job.name = "SeaTunnel"
        spark.executor.instances = 1
        spark.executor.cores = 1
        spark.executor.memory = "1g"
        spark.master = local
      }
      
      source {
        FakeSource {
          row.num = 10
          parallelism = 1
          schema = {
            fields {
              c_map = "map<string, string>"
              c_array = "array<int>"
              c_string = string
              c_boolean = boolean
              c_tinyint = tinyint
              c_smallint = smallint
              c_int = int
              c_bigint = bigint
              c_float = float
              c_double = double
              c_decimal = "decimal(27, 9)"
              c_null = "null"
              c_bytes = bytes
              c_date = date
              c_timestamp = timestamp
            }
          }
          result_table_name = "fake"
        }
      }
      
      transform {
      }
      
      sink {
        SelectDBCloud {
          load-url="ip:port"
          jdbc-url="ip:port"
          cluster-name="SeaTunnel"
          table.identifier="test.TypeTest"
          username="admin"
          password="******"
          selectdb.config {
              file.type="json"
              file.strip_outer_array="false"
          }
        }
      }
```
   * result
   
image
image

### CODE REVIEW

**Comment by TaoZex** [comment]
Please fix code style.
image

You can refer this doc : https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/setup.md

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/StringUtil.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            for (int i = 0; i < len; ++i) {
+                if (!Character.isWhitespace(str.charAt(i))) {
+                    return false;
+                }
+            }
+
+            return true;
+        } else {
+            return true;
+        }
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/StringUtil.java]:
```suggestion
            return true;
```

File: docs/en/connector-v2/sink/SelectDB-Cloud.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+| name                        | type   | required | default value   |
+|-----------------------------|--------|----------|-----------------|
+| load-url                    | string | yes      | -               |
+| jdbc-url                    | string | yes      | -               |
+| cluster-name                | string | yes      | -               |
+| username                    | string | yes      | -               |
+| password                    | string | yes      | -               |
+| table.identifier            | string | yes      | -               |
+| sink.properties             | string | yes      | -               |
```
Comment (TaoZex) [inline docs/en/connector-v2/sink/SelectDB-Cloud.md]: Change config prefix may be better.
Please refer : https://github.com/apache/incubator-seatunnel/pull/3856

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBSinkWriter.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+import org.apache.seatunnel.shade.com.typesafe.config.Config;
+
+import static org.apache.seatunnel.connector.selectdb.sink.writer.LoadConstants.FORMAT_KEY;
+import static org.apache.seatunnel.connector.selectdb.sink.writer.LoadConstants.FIELD_DELIMITER_KEY;
+import static org.apache.seatunnel.connector.selectdb.sink.writer.LoadConstants.LINE_DELIMITER_KEY;
+import static org.apache.seatunnel.connector.selectdb.sink.writer.LoadConstants.LINE_DELIMITER_DEFAULT;
+
+public class SelectDBSinkWriter implements SinkWriter<SeaTunnelRow, SelectDBCommitInfo, SelectDBSinkState> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SelectDBSinkWriter.class);
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBSinkWriter.java]: Please use @slf4j.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/exception/SelectDBConnectorErrorCode.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package org.apache.seatunnel.connector.selectdb.exception;
+
+import org.apache.seatunnel.common.exception.SeaTunnelErrorCode;
+
+public enum SelectDBConnectorErrorCode implements SeaTunnelErrorCode {
+    UPLOAD_FAILED("SelectDB-01", "upload failed "),
+
+    COMMIT_FAILED("SelectDB-02", "commit error with ");
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/exception/SelectDBConnectorErrorCode.java]: Please add in this doc:https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/Error-Quick-Reference-Manual.md

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/ResponseUtil.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * util for handle response.
+ */
+public class ResponseUtil {
+    public static final Pattern LABEL_EXIST_PATTERN =
+            Pattern.compile("errCode = 2, detailMessage = Label \\[(.*)\\] " +
+                    "has already been used, relate to txn \\[(\\d+)\\]");
+//    public static final Pattern COMMITTED_PATTERN =
+//            Pattern.compile("errCode = 2, detailMessage = transaction \\[(\\d+)\\] " +
+//                    "is already \\b(COMMITTED|committed|VISIBLE|visible)\\b, not pre-committed.");
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/ResponseUtil.java]: Remove.

File: docs/en/connector-v2/sink/SelectDB-Cloud.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+### sink.properties [string]
+
+Write property configuration
+CSV Write：
+    sink.properties.file.type='csv' 
+    sink.properties.file.column_separator=',' 
+    sink.properties.file.line_delimiter='\n' 
+JSON Write: 
+    sink.properties.file.type='json' 
+    sink.properties.file.strip_outer_array='false'
```
Comment (TaoZex) [inline docs/en/connector-v2/sink/SelectDB-Cloud.md]: Recommended to support the config prefix, which is a unified configuration format.
https://github.com/apache/incubator-seatunnel/pull/3856/files

File: docs/en/connector-v2/sink/SelectDB-Cloud.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    sink.properties.file.line_delimiter='\n' 
+  }
+}
+```
+
+## Changelog
+
+### next version
+
+- Add SelectDB Cloud Sink Connector
```
Comment (TaoZex) [inline docs/en/connector-v2/sink/SelectDB-Cloud.md]: Please add pr link.You can refer to other documents.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitInfo.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        return Objects.hash(hostPort, clusterName, copySQL);
+    }
+
+    @Override
+    public String toString() {
+        return "SelectDBCommittable{" +
+                "hostPort='" + hostPort + '\'' +
+                ", clusterName='" + clusterName + '\'' +
+                ", copySQL='" + copySQL + '\'' +
+                '}';
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitInfo.java]: It's better to use annotations.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                    break;
+                } else {
+                    log.warn("commit failed, retry again");
+                }
+            }
+        }
+
+        if (!success) {
+            log.error("commit error with status {}, reason {}, response {}", statusCode, reasonPhrase, loadResult);
+            throw new SelectDBConnectorException(SelectDBConnectorErrorCode.COMMIT_FAILED, committable.getCopySQL());
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java]: It is not recommended to use both log.error and throw，and recommended to add the contents of log.error to exception

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordBuffer.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                currentWriteBuffer = null;
+            }
+            if (!isEmpty) {
+                ByteBuffer byteBuffer = writeQueue.take();
+                byteBuffer.flip();
+                checkState(byteBuffer.limit() == 0);
+                readQueue.put(byteBuffer);
+            }
+        } catch (Exception e) {
+            throw new IOException(e);
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordBuffer.java]: Recommended uniform exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                if (StringUtil.isNullOrWhitespaceOnly(loadResult)) {
+                    return null;
+                }
+                log.info("response result {}", loadResult);
+                BaseResponse<HashMap<String, String>> baseResponse = OBJECT_MAPPER.readValue(loadResult, new TypeReference<BaseResponse<HashMap<String, String>>>() {
+                });
+                if (baseResponse.getCode() == 0) {
+                    return baseResponse;
+                } else {
+                    throw new RuntimeException("upload file error: " + baseResponse.getMsg());
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Recommended uniform exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                }
+            }
+            throw new RuntimeException("upload file error: " + response.getStatusLine().toString());
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Recommended uniform exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            String reason = execute.getStatusLine().getReasonPhrase();
+            if (statusCode == HTTP_SUCCESS) {
+                Header location = execute.getFirstHeader("location");
+                String uploadAddress = location.getValue();
+                return uploadAddress;
+            } else {
+                HttpEntity entity = execute.getEntity();
+                String result = entity == null ? null : EntityUtils.toString(entity);
+                log.error("Failed get the redirected address, status {}, reason {}, response {}", statusCode, reason, result);
+                throw new RuntimeException("Could not get the redirected address.");
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Same as above.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+            }
+        }
+    }
+
+    public void close() throws IOException {
+        if (null != httpClient) {
+            try {
+                httpClient.close();
+            } catch (IOException e) {
+                throw new IOException("Closing httpClient failed.", e);
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Recommended uniform exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBSinkWriter.java [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        } catch (Exception ex) {
+            log.error("upload file failed, thread exited:", ex);
+            loadException = ex;
+        }
+    }
+
+    private void checkLoadException() {
+        if (loadException != null) {
+            throw new RuntimeException("error while loading data.", loadException);
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBSinkWriter.java]: Uniform exception

**Review by TaoZex — COMMENTED** [review]
If you do not have a docker environment for e2e testing, add screenshots to prove that the connector is available.

**Comment by hailin0** [comment]
Please add e2e testcase module

references
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-doris-e2e
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java (hunk 9 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java]: I think more information should be displayed in the log here, and exceptions should be thrown more concisely

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java (hunk 13 above)
Comment (zy-kkk) [reply seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Ditto

**Comment by zy-kkk** [comment]
> Please add e2e testcase module
> 
> references https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-doris-e2e https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e

SelectDB Cloud is a cloud native database, and docker cannot be used to start test cases. I will fill in the screenshot of my test results in the description of pull request later

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordBuffer.java (hunk 10 above)
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordBuffer.java]: Unified exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordStream.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    public void endInput() throws IOException {
+        recordBuffer.stopBufferData();
+    }
+
+    @Override
+    public int read(byte[] buff) throws IOException {
+        try {
+            return recordBuffer.read(buff);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordStream.java]: Unified exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordStream.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    public void write(byte[] buff) throws IOException {
+        try {
+            recordBuffer.write(buff);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/RecordStream.java]: Unified exception.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/HttpUtil.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+ * util to build http client.
+ */
+public class HttpUtil {
+    private final HttpClientBuilder httpClientBuilder = HttpClients
+            .custom()
+            .disableRedirectHandling();
+
+    public CloseableHttpClient getHttpClient() {
+        return httpClientBuilder.build();
+    }
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/util/HttpUtil.java]:
```suggestion
public class HttpUtil {
    private HttpUtil() {}

    private final static HttpClientBuilder HTTP_CLIENT_BUILDER = HttpClients
            .custom()
            .disableRedirectHandling();

    public static CloseableHttpClient getHttpClient() {
        return HTTP_CLIENT_BUILDER.build();
    }
```

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/rest/CopyIntoResp.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.apache.seatunnel.connector.selectdb.rest;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.Map;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CopyIntoResp extends BaseResponse {
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/rest/CopyIntoResp.java]:
```suggestion
public class CopyIntoResp extends BaseResponse<Map<String, String>> {
```

**Comment by hailin0** [comment]
> > Please add e2e testcase module
> > references https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-doris-e2e https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e
> 
> SelectDB Cloud is a cloud native database, and docker cannot be used to start test cases. I will fill in the screenshot of my test results in the description of pull request later

Still need to add e2e moduel, but can add annotation `@org.junit.jupiter.api.Disabled`

@zy-kkk

**Comment by zy-kkk** [comment]
> > > Please add e2e testcase module
> > > references https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-doris-e2e https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e
> > 
> > 
> > SelectDB Cloud is a cloud native database, and docker cannot be used to start test cases. I will fill in the screenshot of my test results in the description of pull request later
> 
> Still need to add e2e moduel, but can add annotation `@org.junit.jupiter.api.Disabled`
> 
> @zy-kkk

done

**Review by TaoZex — DISMISSED** [review]
LGTM

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java (hunk 9 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/committer/SelectDBCommitter.java]: Hi @zy-kkk , I aggre with @TaoZex . The reson as the following shown:

When we need throw exception and display error message to user, the best prectice is that not using `log.error()` but printing error message while throwing exception. This kind of practice can make the log message more clearly and improve the user experience.

File: seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java (hunk 13 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-selectdb-cloud/src/main/java/org/apache/seatunnel/connector/selectdb/sink/writer/SelectDBCopyInto.java]: Same as above.

**Review by TyrantLucifer — APPROVED** [review]
LGTM

**Review by hailin0 — APPROVED** [review]

**Pull request merged**